pkg/utils/authcheck/authtype.go

```diff
@@ -0,0 +1,139 @@
+)
+
+type AuthTypes string
+
+type AuthTypeArgs struct {
+	Namespace          string
+	ServiceAccountName string
+	Secret             *corev1.SecretKeySelector
+}
+
+const (
+	AuthenticationCheckUnknownReason = "AuthenticationCheckPending"
+	ControlPlaneNamespace            = "cloud-run-events"
+	BrokerServiceAccountName         = "broker"
+	// Secret option is referring to authentication configuration for secret.
+	// https://cloud.google.com/kubernetes-engine/docs/tutorials/authenticating-to-cloud-platform#importing_credentials_as_a_secret
+	Secret AuthTypes = "secret"
+	// WorkloadIdentityGSA option is referring to authentication configuration for Workload Identity using GSA
+	// https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity
+	WorkloadIdentityGSA AuthTypes = "workload-identity-gsa"
+	WorkloadIdentity    AuthTypes = "workload-identity"
+)
+
+var BrokerSecret = &corev1.SecretKeySelector{
+	LocalObjectReference: corev1.LocalObjectReference{
+		Name: "google-broker-key",
+	},
+	Key: "key.json",
+}
+
+// GetAuthTypeForBrokerCell will get authType for BrokerCell.
+func GetAuthTypeForBrokerCell(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister,
+	secretLister corev1listers.SecretLister, args AuthTypeArgs) (AuthTypes, error) {
+	// For AuthTypeArgs from BrokerCell, ServiceAccountName and Secret will be both presented.
+	// We need to revisit this function after https://github.com/google/knative-gcp/issues/1888 lands,
+	// which will add IdentitySpec to BrokerCell.
+	// For AuthTypeArgs from BrokerCell.
+	authTypeForWorkloadIdentity, workloadIdentityErr := getAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args)
+	authTypeForSecret, secretErr := getAuthTypeForSecret(ctx, secretLister, args)
+	if authTypeForWorkloadIdentity != "" {
+		return authTypeForWorkloadIdentity, nil
+	} else if authTypeForSecret != "" {
+		return authTypeForSecret, nil
+	} else {
+		workloadIdentityError := fmt.Errorf("when checking Kubernetes Service Account %s, got error: %w", args.ServiceAccountName, workloadIdentityErr)
+		secretError := fmt.Errorf("when checking Kubernetes Secret %s, got error: %w", args.Secret.Name, secretErr)
+		return "", fmt.Errorf("authentication is not configured, %s, %s", workloadIdentityError.Error(), secretError.Error())
+	}
+}
+
+// GetAuthTypeForSources will get authType for Sources.
+func GetAuthTypeForSources(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister, args AuthTypeArgs) (AuthTypes, error) {
+	// For AuthTypeArgs from Sources, either ServiceAccountName or Secret will be empty,
+	// because of the IdentitySpec validation from Webhook.
+
+	// For AuthTypeArgs from Sources which has serviceAccountName.
+	if args.ServiceAccountName != "" {
+		authType, err := getAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args)
+		if err != nil {
+			return authType, fmt.Errorf("using Workload Identity for authentication configuration: %w", err)
```

Comment:
```suggestion
			return "", fmt.Errorf("using Workload Identity for authentication configuration: %w", err)
```

Otherwise I'm afraid of an incorrect value being passed through by accident.